simulator/hosts.go

```diff
@@ -0,0 +1,159 @@
+package simulator
+
+import (
+	"io/ioutil"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"sync"
+
+	"github.com/ghodss/yaml"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+)
+
+type Host struct {
+	// Name of this host type.
+	Name string `json:"name"`
+	// parentName is the name of the parent type.
+	ParentName string `json:"parent"`
+	// parent is a pointer to the parent (nil if no parent) used for query
+	// result inheritance.
+	parent *Host
+	//QueryResults maps from regexp pattern to query results that should be
+	//returned.
+	Queries []matcher `json:"queries"`
+
+	// The following members facilitate logging unmatched queries.
+	logger           log.Logger
+	unmatchedMutex   sync.Mutex
+	unmatchedQueries map[string]bool
+}
+
+type matcher struct {
+	// Pattern is a regexp for the query patterns this should match.
+	Pattern regexp.Regexp `json:"pattern"`
+	// Results is the results to return for matched queries
+	Results []map[string]string `json:"results"`
+}
+
+type hostYAML struct {
+	Name       string `json:"name"`
+	ParentName string `json:"parent"`
+	Queries    []struct {
+		Pattern string              `json:"pattern"`
+		Results []map[string]string `json"results"`
+	} `json:"queries"`
+}
+
+func LoadHosts(dir string, logger log.Logger) (map[string]*Host, error) {
+	files, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return nil, errors.Wrap(err, "listing hosts directory")
+	}
+
+	hostMap := map[string]*Host{}
+
+	// Load all files
+	for _, file := range files {
+		if strings.HasSuffix(file.Name(), ".yaml") {
+			path := filepath.Join(dir, file.Name())
+			contents, err := ioutil.ReadFile(path)
+			if err != nil {
+				return nil, errors.Wrapf(err, "reading file %s", path)
+			}
+
+			var h hostYAML
+			err = yaml.Unmarshal(contents, &h)
+			if err != nil {
+				return nil, errors.Wrapf(err, "unmarshal yaml for %s", path)
+			}
+
+			host := Host{
+				Name:             h.Name,
+				ParentName:       h.ParentName,
+				Queries:          []matcher{},
+				unmatchedQueries: make(map[string]bool),
+				logger:           logger,
+			}
+
+			for _, q := range h.Queries {
+				re, err := regexp.Compile(strings.ToLower(q.Pattern))
+				if err != nil {
+					return nil, errors.Wrapf(err, "compile regexp for %s", path)
+				}
+				host.Queries = append(host.Queries, matcher{*re, q.Results})
+			}
+
+			// Check for duplicate host type name
+			if _, exists := hostMap[host.Name]; exists {
+				return nil, errors.Errorf("duplicate host type: %s", host.Name)
+			}
+
+			hostMap[host.Name] = &host
+		}
+	}
+
+	// Link parents
+	for _, host := range hostMap {
+		if host.ParentName == "" {
+			continue
+		}
+
+		parent, ok := hostMap[host.ParentName]
+		if !ok {
+			return nil, errors.Errorf("missing parent named: %s", host.ParentName)
+		}
+		host.parent = parent
+	}
+
+	// TODO check for cycles
```

Comment:
Is this still relevant?

Reply:
Yeah, ideally the data structure that we just assembled is a proper DAG, but we have not actually verified that the graph is acyclic. Ideally the tool would identify the invalid relationship (two files each saying the other is it's parent), but right now that will cause undetermined behavior.

Reply:
We can just try our best to not give it invalid values for `parent` for now though and get back to this some other time IMO.

Reply:
To be precise, if there is a cycle the `RunQuery` call may end up infinite looping if none of the hosts in the cycle have a match for the given query.